pennylane_lightning/lightning_qubit/_adjoint_jacobian.py

```diff
@@ -0,0 +1,344 @@
+        self._qubit_state = qubit_state
+        self._state = qubit_state.state_vector
+        self._dtype = qubit_state.dtype
+        self._jacobian_lightning = (
+            AdjointJacobianC64() if self._dtype == np.complex64 else AdjointJacobianC128()
+        )
+        self._create_ops_list_lightning = (
+            create_ops_listC64 if self._dtype == np.complex64 else create_ops_listC128
+        )
+        self._batch_obs = batch_obs
+
+    @property
+    def qubit_state(self):
+        """Returns a handle to the LightningStateVector class."""
+        return self._qubit_state
+
+    @property
+    def state(self):
+        """Returns a handle to the Lightning internal data class."""
+        return self._state
+
+    @property
+    def dtype(self):
+        """Returns the simulation data type."""
+        return self._dtype
+
+    @staticmethod
+    def _check_adjdiff_supported_measurements(
+        measurements: List[MeasurementProcess],
+    ):
+        """Check whether given list of measurement is supported by adjoint_differentiation.
+
+        Args:
+            measurements (List[MeasurementProcess]): a list of measurement processes to check.
+
+        Returns:
+            Expectation or State: a common return type of measurements.
+        """
+        if not measurements:
+            return None
+
+        if len(measurements) == 1 and measurements[0].return_type is State:
+            return State
+
+        # Now the return_type of measurement processes must be expectation
+        if any(measurement.return_type is not Expectation for measurement in measurements):
+            raise QuantumFunctionError(
+                "Adjoint differentiation method does not support expectation return type "
+                "mixed with other return types"
+            )
+
+        for measurement in measurements:
+            if isinstance(measurement.obs, Tensor):
+                if any(isinstance(obs, Projector) for obs in measurement.obs.non_identity_obs):
+                    raise QuantumFunctionError(
+                        "Adjoint differentiation method does "
+                        "not support the Projector observable"
+                    )
+            elif isinstance(measurement.obs, Projector):
+                raise QuantumFunctionError(
+                    "Adjoint differentiation method does not support the Projector observable"
+                )
+        return Expectation
+
+    @staticmethod
+    def _check_adjdiff_supported_operations(operations):
+        """Check Lightning adjoint differentiation method support for a tape.
+
+        Raise ``QuantumFunctionError`` if ``tape`` contains not supported measurements,
+        observables, or operations by the Lightning adjoint differentiation method.
+
+        Args:
+            tape (.QuantumTape): quantum tape to differentiate.
+        """
+        for operation in operations:
+            if operation.num_params > 1 and not isinstance(operation, Rot):
```

Comment:
We should support `qml.QubitUnitary` since [this story](https://app.shortcut.com/xanaduai/story/38319/the-adjoint-method-in-lightning-supports-arbitrary-non-controlled-operations). We are currently limited to 1- and 2-qubit unitaries. The limitation currently comes from the ability to perform `qml.QubitUnitary` decomposition beyond that size.